File: notes/0501/0501.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
        
    def findMode(self, root: TreeNode) -> List[int]:
        count = {}
        def dfs(TreeNode):
            if TreeNode:
                if TreeNode.val in count:
                    count[TreeNode.val] += 1
                else:
                    count[TreeNode.val] = 1
                dfs(TreeNode.left)
                dfs(TreeNode.right)
        dfs(root)
        max_ = 0
        res = []
        for key, value in count.items():
            if value > max_:
                max_ = value
        for key, value in count.items():
            if max_ == value:
                res.append(key)
        return res

File: notes/0501/0501.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> findMode(TreeNode* root) {
        inorder(root);
        return ans;
    }
private:
    int cnt = 0;
    int max_cnt = 0;
    vector<int> ans;
    int pre = -1;
    void inorder(TreeNode* root) {
        if(!root) return;
        inorder(root->left);
        if (root->val == pre) {
            cnt += 1;
        }
        else {
            cnt = 0;
            pre = root->val;
        }
        if (cnt > max_cnt) {
            max_cnt = cnt;
            ans.clear();
        }
        if (max_cnt == cnt) {
            ans.push_back(root->val);
        }
        inorder(root->right);
    }
};
